File: DuDoong-Domain/src/main/java/band/gosrock/domain/domains/ticket_item/domain/OptionGroup.java
package band.gosrock.domain.domains.ticket_item.domain;

import static band.gosrock.common.consts.DuDoongStatic.KR_NO;
import static band.gosrock.common.consts.DuDoongStatic.KR_YES;
import static band.gosrock.domain.common.vo.Money.ZERO;
import static band.gosrock.domain.domains.ticket_item.domain.OptionGroupType.*;

import band.gosrock.domain.common.vo.Money;
import band.gosrock.domain.domains.ticket_item.exception.ForbiddenOptionGroupDeleteException;
import band.gosrock.domain.domains.ticket_item.exception.InvalidOptionGroupException;
import java.util.ArrayList;
import java.util.List;
import javax.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.ColumnDefault;

@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Entity(name = "tbl_option_group")
public class OptionGroup {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "option_group_id")
    private Long id;

    private Long eventId;

    // 옵션 그룹 응답 형식
    @Enumerated(EnumType.STRING)
    private OptionGroupType type;

    // 옵션 그룹 이름
    private String name;

    // 옵션 그룹 설명
    private String description;

    // 필수 응답 여부
    private Boolean isEssential;

    // 상태
    @Enumerated(EnumType.STRING)
    @ColumnDefault(value = "'VALID'")
    private OptionGroupStatus optionGroupStatus = OptionGroupStatus.VALID;

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "optionGroup")
    private final List<Option> options = new ArrayList<>();

    @Builder
    public OptionGroup(
            Long eventId,
            OptionGroupType type,
            String name,
            String description,
            Boolean isEssential,
            List<Option> options) {
        this.eventId = eventId;
        this.type = type;
        this.name = name;
        this.description = description;
        this.isEssential = isEssential;

        this.options.addAll(options);
        options.forEach(option -> option.setOptionGroup(this));
    }

    public void validateEventId(Long eventId) {
        if (!this.getEventId().equals(eventId)) {
            throw InvalidOptionGroupException.EXCEPTION;
        }
    }

    public Boolean hasApplication(List<TicketItem> ticketItems) {
        return ticketItems.stream()
                .map(ticketItem -> ticketItem.hasItemOptionGroup(this.id))
                .reduce(Boolean.FALSE, Boolean::logicalOr);
    }

    public OptionGroup createTicketOption(Money additionalPrice) {
        OptionGroupType type = this.getType();
        if (type == TRUE_FALSE) {
            this.options.add(Option.create(KR_YES, additionalPrice, this));
            this.options.add(Option.create(KR_NO, ZERO, this));
        } else if (type == SUBJECTIVE) {
            this.options.add(Option.create("", ZERO, this));
        }
        return this;
    }

    public void softDeleteOptionGroup(List<TicketItem> ticketItems) {
        // 적용된 옵션은 삭제 불가
        if (this.hasApplication(ticketItems)) {
            throw ForbiddenOptionGroupDeleteException.EXCEPTION;
        }
        this.optionGroupStatus = OptionGroupStatus.DELETED;
    }
}

File: DuDoong-Domain/src/main/kotlin/band/gosrock/domain/domains/ticket_item/domain/OptionGroup.kt
package band.gosrock.domain.domains.ticket_item.domain

import band.gosrock.common.consts.DuDoongStatic.KR_NO
import band.gosrock.common.consts.DuDoongStatic.KR_YES
import band.gosrock.domain.common.vo.Money
import band.gosrock.domain.domains.ticket_item.domain.OptionGroupType.SUBJECTIVE
import band.gosrock.domain.domains.ticket_item.domain.OptionGroupType.TRUE_FALSE
import band.gosrock.domain.domains.ticket_item.exception.ForbiddenOptionGroupDeleteException
import band.gosrock.domain.domains.ticket_item.exception.InvalidOptionGroupException
import band.gosrock.domain.domains.ticket_item.exception.InvalidOptionPriceException
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.OneToMany
import org.hibernate.annotations.ColumnDefault

@Entity(name = "tbl_option_group")
class OptionGroup(
    var eventId: Long? = null,
    // 옵션 그룹 응답 형식
    @Enumerated(EnumType.STRING)
    var type: OptionGroupType? = null,
    // 옵션 그룹 이름
    var name: String? = null,
    // 옵션 그룹 설명
    var description: String? = null,
    // 필수 응답 여부
    var isEssential: Boolean? = null,
    initialOptions: List<Option> = emptyList(),
) {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "option_group_id")
    var id: Long? = null
        protected set

    // 상태
    @Enumerated(EnumType.STRING)
    @ColumnDefault(value = "'VALID'")
    var optionGroupStatus: OptionGroupStatus = OptionGroupStatus.VALID
        protected set

    @OneToMany(cascade = [CascadeType.ALL], mappedBy = "optionGroup")
    val options: MutableList<Option> = mutableListOf()

    init {
        if (initialOptions.isNotEmpty()) {
            this.options.addAll(initialOptions)
            initialOptions.forEach { it.updateOptionGroup(this) }
        }
    }

    fun validateEventId(eventId: Long) {
        if (this.eventId != eventId) throw InvalidOptionGroupException.EXCEPTION
    }

    fun hasApplication(ticketItems: List<TicketItem>): Boolean =
        ticketItems.any { it.hasItemOptionGroup(this.id!!) }

    fun createTicketOption(additionalPrice: Money): OptionGroup {
        when (type) {
            TRUE_FALSE -> {
                if (additionalPrice.isLessThan(Money.ZERO)) throw InvalidOptionPriceException.EXCEPTION
                this.options.add(Option.create(KR_YES, additionalPrice, this))
                this.options.add(Option.create(KR_NO, Money.ZERO, this))
            }
            SUBJECTIVE -> {
                this.options.add(Option.create("", Money.ZERO, this))
            }
            else -> {}
        }
        return this
    }

    fun softDeleteOptionGroup(ticketItems: List<TicketItem>) {
        // 적용된 옵션은 삭제 불가
        if (this.hasApplication(ticketItems)) throw ForbiddenOptionGroupDeleteException.EXCEPTION
        this.optionGroupStatus = OptionGroupStatus.DELETED
    }
}
